VDA-StorageNode-Registry/contracts/StorageNodeRegistry.sol

```diff
@@ -0,0 +1,1098 @@
+                nodeId = ids.at(i);
+                if (_nodeUnregisterTime[nodeId] == 0) {
+                    nodeList[index] = _nodeMap[nodeId];
+                    ++index;
+                }
+                unchecked { ++i; }
+            }
+        }
+
+        return nodeList;
+    }
+
+    /**
+     * @dev see { IStorageNodeRegistry }
+     */
+    function getNodesByCountry(string calldata countryCode) external view override returns(StorageNode[] memory) {
+        return filterActiveStorageNodes(_countryNodeIds[countryCode]);
+    }
+
+    /**
+     * @dev see { IStorageNodeRegistry }
+     */
+    function getNodesByRegion(string calldata regionCode) external view override returns(StorageNode[] memory) {
+        return filterActiveStorageNodes(_regionNodeIds[regionCode]);
+    }
+
+    /**
+     * @dev see { IStorageNodeRegistry }
+     */
+    function isStakingRequired() external view returns(bool) {
+        return _slotInfo.isStakingRequired;
+    }
+
+    /**
+     * @dev see { IStorageNodeRegistry }
+     */
+    function setStakingRequired(bool isRequired) external payable onlyOwner override {
+        if (isRequired == _slotInfo.isStakingRequired) {
+            revert InvalidValue();
+        }
+
+        _slotInfo.isStakingRequired = isRequired;
+        emit UpdateStakingRequired(isRequired);
+    }
+
+    /**
+     * @dev see { IStorageNodeRegistry }
+     */
+    function getStakePerSlot() external view returns(uint) {
+        return _slotInfo.STAKE_PER_SLOT;
+    }
+
+    /**
+     * @dev see { IStorageNodeRegistry }
+     */
+    function updateStakePerSlot(uint newVal) external payable onlyOwner override {
+        if (newVal == 0 || newVal == _slotInfo.STAKE_PER_SLOT) {
+            revert InvalidValue();
+        }
+
+        _slotInfo.STAKE_PER_SLOT = newVal;
+        emit UpdateStakePerSlot(newVal);
+    }
+
+    /**
+     * @dev see { IStorageNodeRegistry }
+     */
+    function getSlotCountRange() external view returns(uint, uint) {
+        return (_slotInfo.MIN_SLOTS, _slotInfo.MAX_SLOTS);
+    }
+
+    /**
+     * @dev see { IStorageNodeRegistry }
+     */
+    function updateMinSlotCount(uint minSlots) external payable onlyOwner override {
+        if (minSlots == 0 || minSlots == _slotInfo.MIN_SLOTS || minSlots > _slotInfo.MAX_SLOTS) {
+            revert InvalidValue();
+        }
+
+        _slotInfo.MIN_SLOTS = minSlots;
+        emit UpdateMinSlotCount(minSlots);
+    }
+
+    /**
+     * @dev see { IStorageNodeRegistry }
+     */
+    function updateMaxSlotCount(uint maxSlots) external payable onlyOwner override {
+        if (maxSlots == 0 || maxSlots == _slotInfo.MAX_SLOTS || maxSlots < _slotInfo.MIN_SLOTS) {
+            revert InvalidValue();
+        }
+
+        _slotInfo.MAX_SLOTS = maxSlots;
+        emit UpdateMaxSlotCount(maxSlots);
+    }
+
+    /**
+     * @dev see { IStorageNodeRegistry }
+     */
+    function updateTokenAddress(address newTokenAddress) external payable onlyOwner {
```

Comment:
We shouldn't every cahnge the token address.. so is this a security risk?

Reply:
This might be a security risk.
Users don't like to change the token address.
Also, there are many users and staked tokens, it might be a bit complicated for us to change the token. Of course we can change the token address simply. After that, we should convert current staked tokens to the new token. This'd be so complicated.

Reply:
Ok. Let's remove the ability to change the token.

Reply:
Ok